src/main/java/org/commcare/xml/bulk/BulkProcessingCaseXmlParser.java

```diff
@@ -0,0 +1,308 @@
+package org.commcare.xml.bulk;
+
+import org.commcare.cases.model.Case;
+import org.commcare.cases.model.CaseIndex;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.utils.DateUtils;
+import org.javarosa.core.util.externalizable.SerializationLimitationException;
+import org.javarosa.xml.util.ActionableInvalidStructureException;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.kxml2.io.KXmlParser;
+
+import java.util.Date;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * A parser which is capable of processing CaseXML transactions in a bulk format.
+ *
+ * This parser needs an implementation which can perform the "bulk" steps efficiently on the current
+ * platform.
+ *
+ * It should be a drop-in replacemenet for the CaseXmlParser in the ways that it is used
+ *
+ * Created by ctsims on 3/14/2017.
+ */
+public abstract class BulkProcessingCaseXmlParser extends BulkElementParser<Case> {
+
+    public BulkProcessingCaseXmlParser(KXmlParser parser) {
+        super(parser);
+    }
+
+    @Override
+    protected void requestModelReadsForElement(TreeElement bufferedTreeElement, Set<String> currentBulkReadSet) {
+        String caseId = bufferedTreeElement.getAttributeValue(null, "case_id");
+        currentBulkReadSet.add(caseId);
+    }
+
+    @Override
+    protected void preParseValidate() throws InvalidStructureException {
+        checkNode("case");
+    }
+
+    @Override
+    protected void processBufferedElement(TreeElement bufferedTreeElement, Map<String, Case> currentOperatingSet, LinkedHashMap<String, Case> writeLog) throws InvalidStructureException {
```

Comment:
Hey,

It is identical except for the fact that one parses a streamed XML element and the other parses our intermediate buffered tree element object, and that makes the parsing very different structurally.

I looked at how to try to merge the two and it was super, super ugly, since you "consume" elements in the other parser, and in this one you query them from an in-memory object. The other parser is also capable, for instance, of reporting errors based on the current location in the parse, which this can't do.

I'll make a ticket to try again to merge the implementations, but it is going to be super ugly no matter what due to those differences. It likely makes sense once we have confidence in this parser to make the "default" parser this parser with a "trigger" buffer of 1 rather than maintaining both.